File: Python/rules/speed_rule.py
from rules.rule import Rule
from model.violation import Violation

class SpeedRule(Rule):
    def __init__(self, max_speed_by_car_type, fee):
        self.max_speed_by_car_type = max_speed_by_car_type
        self.fee = fee

    def evaluate(self, observation):
        max_speed = self.max_speed_by_car_type.get(observation.get_car_type())
        if max_speed is not None and observation.get_speed() > max_speed:
            description = "speed of " + str(observation.get_speed()) + " exceeded max allowed " + str(max_speed)
            return Violation(description, self.fee)
        return None

File: JAVA/rules/SpeedRule.java
package rules;
import model.CarObservation;
import model.CarType;
import model.Violation;
import java.util.Map;
import java.util.Optional;

public class SpeedRule implements Rule {
    private final Map<CarType, Integer> maxSpeedByCarType;
    private final double fee;
    public SpeedRule(Map<CarType, Integer> maxSpeedByCarType, double fee) {
        this.maxSpeedByCarType = maxSpeedByCarType;
        this.fee = fee;
    }
    @Override
    public Optional<Violation> evaluate(CarObservation observation) {
        Integer maxSpeed = maxSpeedByCarType.get(observation.getCarType());
        if (maxSpeed != null && observation.getSpeed() > maxSpeed) {
            String description = "speed of " + observation.getSpeed()+ " exceeded max allowed " + maxSpeed;
            return Optional.of(new Violation(description, fee));
        }
        return Optional.empty();
    }
}